core/platform/cloud_translate/cloud_translate_emulator.py

```diff
@@ -0,0 +1,107 @@
+
+import python_utils
+import utils
+
+
+class CloudTranslateEmulator(python_utils.OBJECT):
+    """The emulator mocks the translate_text function from the Cloud Translate
+    API. This emulator can be used in backend testing, or a local dev
+    environment without access to the Cloud Translate API. Expected responses
+    must be passed in before using this emulator for testing. See
+    PREGENERATED_TRANSLATIONS above for some prepopulated responses.
+    """
+
+    def __init__(self):
+        """Initializes the emulator with no expected responses."""
+        self.DEFAULT_RESPONSE = (
+            'Default translation for emulator mode. (See core/platform/'
+            'cloud_translate/cloud_translate_emulator.py for details)')
+
+        # Pre-generated translations for the following phrases:
+        # ('hello world', 'CONTINUE', 'Please continue.', 'Correct!')
+        self.PREGENERATED_TRANSLATIONS = {
+            (u'en', u'pt', u'hello world'): u'Ol\xe1 Mundo',
+            (u'en', u'pt', u'CONTINUE'): u'PROSSEGUIR',
+            (u'en', u'es', u'Please continue.'): u'Por favor continua.',
+            (u'en', u'fr', u'CONTINUE'): u'CONTINUEZ',
+            (u'en', u'fr', u'Please continue.'):
+                u'Continuez s&#39;il vous pla\xeet.',
+            (u'en', u'es', u'CONTINUE'): u'SEGUIR',
+            (u'en', u'zh', u'hello world'): u'\u4f60\u597d\u4e16\u754c',
+            (u'en', u'es', u'Correct!'): u'\xa1Correcto!',
+            (u'en', u'zh', u'Correct!'): u'\u6b63\u786e\u7684\uff01',
+            (u'en', u'zh', u'CONTINUE'): u'\u7ee7\u7eed',
+            (u'en', u'zh', u'Please continue.'): u'\u8bf7\u7ee7\u7eed\u3002',
+            (u'en', u'fr', u'Correct!'): u'Correct!',
+            (u'en', u'pt', u'Correct!'): u'Correto!',
+            (u'en', u'es', u'hello world'): u'Hola Mundo',
+            (u'en', u'pt', u'Please continue.'): u'Por favor continue.',
+            (u'en', u'fr', u'hello world'): u'Bonjour le monde'
+            }
+        self.expected_responses = self.PREGENERATED_TRANSLATIONS
+
+    def translate(self, text, source_language_code, target_language_code):
+        """Returns the saved expected response for a given input. If no
+        response exists for the given input, returns a default response.
+
+        For more information on ISO 639-1 see:
+            https://www.w3schools.com/tags/ref_language_codes.asp
+
+        Args:
+            text: str. The text to be translated.
+            source_language_code: str. A allowlisted ISO 639-1 language code.
+            target_language_code: str. A allowlisted ISO 639-1 language code.
+
+        Raises:
+            ValueError. Invalid source language code.
+            ValueError. Invalid target language code.
+
+        Returns:
+            str. The translated text.
+        """
+        if not utils.is_valid_language_code(source_language_code):
+            raise ValueError('Invalid language code: %s' % source_language_code)
```

Comment:
Please also add source/target to these error messages so it's obvious which one was wrong when the exception is raised.

Reply:
Done.